src/main/scala/rise/core/Expr.scala

```diff
@@ -1,46 +1,81 @@
 package rise.core
 
+import semantics._
 import rise.core.types._
-import rise.core.DSL.Type.TypeEqual
 import rise.core.ShowRise._
+import util.PatternMatching
+
+object alphaEquiv {
+  type TYPEEQ = Type => Type => Boolean
+
+  val equiv : TYPEEQ => Expr => Expr => Boolean = typeEq => a => b => {
+    // Make the match exhaustive
+    val and = PatternMatching.matchWithDefault(b, false)
+    typeEq(a.t)(b.t) && (a match {
+      case Identifier(na) => and { case Identifier(nb) => na == nb }
+      case Literal(da) => and { case Literal(db) => da == db }
+      case a : Primitive => and { case b : Primitive => a.name == b.name }
+      // Application compares structurally
+      case App(fa, ea) => and { case App(fb, eb) => equiv(typeEq)(fa)(fb) && equiv(typeEq)(ea)(eb) }
+      case DepApp(fa, xa) => and { case DepApp(fb, xb) => equiv(typeEq)(fa)(fb) && xa == xb }
+      // Abstraction compares after substitution
+      case Lambda(xa, ta) => and { case other@Lambda(xb, _) =>
+        typeEq(xa.t)(xb.t) && equiv(typeEq)(ta)(typedLifting.liftFunExpr(other).value(xa)) }
+      case DepLambda(xa, ea) => and { case other@DepLambda(xb, _) =>
+        val and = PatternMatching.matchWithDefault(xb, false)
+        xa match {
+          case n: NatIdentifier => and { case _: NatIdentifier =>
+            equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatKind](other).value(n)) }
+          case dt: DataTypeIdentifier => and { case _: DataTypeIdentifier =>
+            equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[DataKind](other).value(dt)) }
+          case addr: AddressSpaceIdentifier => and { case _: AddressSpaceIdentifier =>
+            equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[AddressSpaceKind](other).value(addr)) }
+          case n2n: NatToNatIdentifier => and { case _: NatToNatIdentifier =>
+            equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToNatKind](other).value(n2n)) }
+          case n2d: NatToDataIdentifier => and { case _: NatToDataIdentifier =>
+            equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToDataKind](other).value(n2d)) }
+      }}
+    })
+  }
```

Comment:
it would be good to avoid substitution for expressions as well, you will then have to communicate the environment for type equality.

Reply:
Agreed! It's the next item on my todo ^^